### ISSUE TITLE
`prisma format` ends the file with a single CRLF on windows

### ISSUE THREAD

**Issue by binary64** [prisma/prisma#8548]
### Bug description

On windows, I run `prisma format` and note the unusual file ending. The lines are all LF, but the very last line is CRLF.

This causes issue on my Linux CI where it formats it ending in LF's only, causing a diff to occur and the build to fail.

### How to reproduce

1. On windows do prisma format
2. Open in HxD or similar
3. See attached:
 
![image](https://user-images.githubusercontent.com/1680627/127931864-fcd66391-9b31-4914-8553-f6c71b3a6fad.png)

### Expected behavior

LF's only - no CR's to exist at all

### Environment & setup

Node 16
Windows/Linux

### Prisma Version

```
$ C:\p\pab\monorepo\node_modules\.bin\prisma -v
prisma               : 2.26.0
@prisma/client       : 2.28.0
Current platform     : windows
Query Engine         : query-engine 9b816b3aa13cc270074f172f30d6eda8a8ce867d (at node_modules\@prisma\engines\query-engine-windows.exe)
Migration Engine     : migration-engine-cli 9b816b3aa13cc270074f172f30d6eda8a8ce867d (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine : introspection-core 9b816b3aa13cc270074f172f30d6eda8a8ce867d (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary        : prisma-fmt 9b816b3aa13cc270074f172f30d6eda8a8ce867d (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash : 9b816b3aa13cc270074f172f30d6eda8a8ce867d
Studio               : 0.408.0
```

**Comment by Miladiir** [comment]
I just ran into the same problem. I check out files as-is, so without git's auto-clrf setting, even on windows.

**Comment by DanielZavala11** [comment]
Hey @janpio, Guillermo (@GmoZen) and I are working on a fix for this bug as part of our CodeDay project (https://labs.codeday.org/).
I'm wondering if we can fix this by just changing the os.EOL line in this file to \n,  that way the file would consistently have the same line endings on every line, as opposed to having the platform-specific line ending on the last line?
https://github.com/prisma/prisma/blob/96dc3444b1439e3ab166a69c8290aec06c61bfba/packages/cli/src/Format.ts#L65
Or is the goal here to have every line have the correct platform-specific EOL? If so, would this require a change in the prisma engines Rust code, somewhere here?
https://github.com/prisma/prisma-engines/blob/main/prisma-fmt/src/format.rs
Thank you!

**Comment by Miladiir** [comment]
Would it be enough to get one sample of a line ending from the file and then using that instead of os.EOL?

**Comment by janpio** [comment]
@DanielZavala11 Figuring that out is part of the work here - I do not know what the _real_ expected and correct behavior would be. That depends very much on what problems the current behaviour is creating, and if the new would avoid these. So right now we would need to _understand_ further (optimally with a reproduction(before we can fix.

**Comment by GmoZen** [comment]
Thanks for the advice Jan!
@Miladiir, can you describe a bit more the problem this causes for you? Is the issue that you (1) are editing a file on one OS (let's say Windows), and you push that, (2) another project member edits the same file on another OS (let's say Mac) and edits it (changing one line) and runs `prisma format`, (3) git then shows that ALL lines have changed, because the format command changed all the line endings?
If so, we're thinking that maybe your suggestion of sampling a line ending from the file is a good solution. Or alternatively always just sticking to one type of line ending (let's say just `LF`) but I'm not sure if that has other consequences. We'll try to research that a bit more.

**Comment by janpio** [comment]
This all get more complicate when a team setting and Git come into play:
1. You might have two developers working on Linux and Windows - one creates the file initially, the other later formats it. Which line ending to use?
2. Git has configurable behavior on Windows to locally change the line endings to the Windows default, but read and write the Linux default instead to avoid exactly situations like listed in 1.
3. If an existing file already has different line endings when you start, things get even messier.

**Comment by Nike682631** [comment]
@janpio  Is this issue available to work on?

**Comment by janpio** [comment]
Generally yes, but we have a previous attempt here: https://github.com/prisma/prisma/pull/12401 I am not sure why that did not progress further. 

If you want to take another shot, a first step should to be to confirm and understand the problem and offer an overview of the problem and a potential solution here in the issue.

**Comment by Jolg42** [comment]
With @tomhoule we think this could be already fixed, so @Nike682631, check if you can reproduce with the latest Prisma version or not first, if it's fixed we can close this 😄

**Comment by tomhoule** [comment]
I stumbled on where this is happening by chance: https://github.com/prisma/prisma/blob/a1b1f3cb3780c4214f125911791225a8f7d48685/packages/cli/src/Format.ts#L82 — it looks like the CLI is adding this.

**Comment by fvckDesa** [comment]
This issue is already solved

**Comment by petradonka** [comment]
Thanks for looking into this @fvckDesa! 🙌 

We can extend the scope of this issue to allow you to work on the bounty and create a meaningful contribution, also considering the work you've already put into it. We suggest the following:
- confirming via our tests that all operating systems indeed use the same line ending consistently
- cover other commands besides `prisma format`, e.g `prisma init` or `prisma migrate diff`, which also create schema files

@jharrell can help you if you have any questions about the details.

**Comment by fvckDesa** [comment]
@petradonka It's fine for me, I start working on it as soon as I finish with another commitment.
/attempt https://github.com/prisma/prisma/issues/8548

**Comment by fvckDesa** [comment]
I can't reattempt the bounty, in the case I can't work on it I'll leave a comment

**Comment by opiredev** [comment]
@nabby27 created a $50.00 reward using [Opire](https://opire.dev) 
 <details><summary>How to earn this reward?</summary>

Since this project does not have Opire bot installed yet 😞 you need to go to [Opire](https://app.opire.dev) and claim the rewards through your programmer's dashboard once you have your PR ready 💪</details>

**Comment by jharrell** [comment]
Hi there @nabby27,

we are not accepting bounties at this time. Thank you for your interest, however!

**Comment by nabby27** [comment]
Hi @jharrell, do you not accept rewards for this issue or for any issue? Just to know if I could contribute financially to any issue (even if it is a small one).

**Comment by jharrell** [comment]
@nabby27 👋 

TL;DR - we don't accept rewards for any issues. 


We trialed a few different bounty systems with a small set of issues, you happened to find one of them 😁 We're still thinking about this though, just want to find the best way to approach it in order to do right by our community.

**Comment by danielcoderman** [comment]
Just curious, why hasn't this issue been closed yet? I'm new to contributing, and I read through here and based off the comments it appears to have been addressed already.

**Comment by tysoncung** [comment]
I'm interested in working on this issue. Could you provide more context about what you're looking for? Any additional details about requirements or constraints would be helpful.

**Comment by dohrisalim** [comment]
I can fix this cross-platform line ending inconsistency in the prisma format command. I understand the problem: mixed line endings (LF for most lines, CRLF for the last) cause CI/CD failures on Linux. I'll ensure the formatter outputs consistent LF line endings across all platforms, preventing unexpected formatting changes when code moves between Windows and Unix environments.

**Comment by Iceshen87** [comment]
✅ PR Created: https://github.com/prisma/prisma/pull/29299

**Fix:**
- Normalize line endings before writing to disk
- Prevents extra CRLF on Windows
- Uses os.EOL for platform-specific line endings

**Testing:**
- Windows: No extra CRLF
- Unix: LF preserved
- `prisma format --check` works correctly

Ready for review! 🚀

**Comment by Iceshen87** [comment]
## Review Response - PR #29299

Thanks for the 2 review comments!

1. ✅ Line ending normalization uses os.EOL for platform-specific behavior
2. ✅ Tested on Windows (no extra CRLF) and Unix (LF preserved)

Ready for re-review! 🚀

**Comment by chengyixu** [comment]
Hi! I'd like to work on this CRLF issue.

**Approach:** The root cause is likely in Prisma's formatter (written in Rust) where the file write on Windows uses platform-native line endings for the final newline but LF for content. The fix would be to normalize the final byte written to LF explicitly in the format output writer, regardless of OS. I'd trace it through the prisma-fmt crate, find where the output string is written to disk, and add an explicit LF normalization for the trailing newline.

**ETA:** 2-3 days
**Background:** I maintain 35+ npm packages and work extensively in this ecosystem. Familiar with Rust's std::io::Write and cross-platform line ending handling.

Happy to discuss before starting. — @chengyixu

**Comment by chengyixu** [comment]
Hi! I'd like to tackle this.

**Approach:** The issue is that `prisma format` writes the last line with CRLF on Windows while keeping all other lines as LF. I'll locate the file-write path in the Prisma CLI formatter (TypeScript), normalize line endings before writing by stripping all `\r` characters and ensuring consistent LF output, then add a Windows-specific CI test to verify the fix.
**Timeline:** 2-3 days
**Background:** Node.js/TypeScript developer, 35+ npm packages published. Experience with cross-platform file encoding and CI pipeline setups.

Ready to start immediately. — @chengyixu

**Comment by lky141319** [comment]
Claiming bounty. 
Wallet: 0xd2268DEDca16b08aF4bb6Ad20cBa877D10cf8Ac6
GitHub: @lky141319

**Comment by SevInf** [comment]
To anybody handling this issue to their agent: this repo is the wrong target for your pull request. Fix should go into prisma-format code itself. Good starting point is [here](https://github.com/prisma/prisma-engines/blob/main/psl/schema-ast/src/reformat.rs).
We will not accept pull requests that target CLI, split prisma-fmt output and re-join it with different line-ending unconditionally. We will also not accept pull-request that unconditionally replaces CR-LF ending with LF for all Windows users. We will accept one of the following:
- setting to switch between CR-LF and LF with CR-LF being a default on Windows and LF everywhere else
- preserving original line endings in the file (what #29373 did, but in Rust codebase)

**Comment by SevInf** [comment]
One more thing: bounty on this is not active. It haven't been active for many years. I don't know why your agent thinks it is, but we will not be paying for this fix.

**Comment by baronguyen001** [comment]
Opened PR https://github.com/prisma/prisma-engines/pull/5814 for this issue.
